### ISSUE TITLE
[java] EmptyTryBlock: false positive for empty try-with-resource

### ISSUE THREAD

**Issue by dreniers** [pmd/pmd#432]
**Rule Set:**

rulesets/java/empty.xml/EmptyTryBlock

**Description:**

The rule says "Avoid empty try blocks - what's the point?". I believe an exception should be made for empty try-with-resource blocks, as this is a valid use case. 

**Code Sample demonstrating the issue:**

```
class X {
	void method() {
		try {
			// true positive
		} catch (Exception e) {
		}

		try (ClientResponse response = execute(() -> target.request(mediaTypes).delete(), DELETE, new ExpectedResponse(status, required))) {
			// false positive
		}
	}
}
```

**Running PMD through:** *CLI*

```
pmd.bat -d X.java -f xml -R rulesets/java/empty.xml/EmptyTryBlock
```

**Comment by dreniers** [comment]
As as fix, I propose to change:
```
//TryStatement/Block[1][count(*) = 0]
```
into
```
 //TryStatement[not(ResourceSpecification)]/Block[1][count(*) = 0]
```
What do you think?

**Comment by adangel** [comment]
@dreniers Thanks for the suggestion!

Yes, it sounds reasonable to make an exception for an empty try-with-resource, since
* you need define a AutoClosable resource (the parenthesis can't be empty)
* java will then close the resource for you
* any exceptions raised either need to be caught in the same try with an additional catch or declared to be thrown by the method.

Just to double check: Which API is `ClientResponse` exactly? I assume, it implements the AutoClosable interface, but what exactly is closed? I looks like a REST request, but unless the connection is pooled/pipelined, it is already closed.

**Comment by dreniers** [comment]
@adangel Sorry for the late reply, I was on vacation.

> Just to double check: Which API is ClientResponse exactly? I assume, it implements the AutoClosable interface, but what exactly is closed?

I am not sure, our customer only disclosed this code snippet. 

Anyway, thanks for the fix!
